cleanlab/dataset.py

```diff
@@ -0,0 +1,150 @@
+# Copyright (C) 2017-2022  Cleanlab Inc.
+# This file is part of cleanlab.
+#
+# cleanlab is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as published
+# by the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# cleanlab is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with cleanlab.  If not, see <https://www.gnu.org/licenses/>.
+
+
+"""Dataset Module
+Supports dataset-level and class-level automated quality, including finding which classes should
+be merged with other classes in your dataset, which classes should be removed,
+and which classes have the best label accuracy, and which have the worst labels.
+"""
+
+import numpy as np
+from count import estimate_joint, compute_confident_joint
+
+
+def find_similar_classes(
+    joint=None,
+    asymmetric=True,
+    *,
+    labels=None,
+    pred_probs=None,
+    confident_joint=None,
+    multi_label=False,
+):
+    """Ranks all pairs of classes by their similarity and returns the ordered pairs as well as
+    a similarity score between 0 and 1. Use this function to determine which classes in a dataset
+    should be merged.
+
+    The method uses the joint distribution of noisy and true labels to compute ontological issues
+    via the approach published in (Northcutt, Jiang, Chuang (2021) JAIR).
+    paper: https://arxiv.org/abs/1911.00068
+
+    This method works by providing any one (and only one) of the following inputs:
+    1. confident_joint
+    2. joint
+    3. labels and pred_probs
+    but it is not necessary to provide a combination, e.g. the labels and pred_probs and joint.
+
+    For parameter info, see the docstring of `count.estimate_joint`
+
+    Parameters
+    ----------
+    See `count.estimate_joint` docstring.
+
+    asymmetric : bool (default: True)
+      If `asymmetric==True`, includes both pairs (class1, class2) and (class2, class1). Use this for
+      finding "is a" relationships where for example "class1 is a class2".
+      If `asymmetric==False`, the pair (class1, class2) will onyl be returned once and order is
+      arbitrary (internally this is just summing score(class1, class2) + score(class2, class1)).
+
+    Returns
```

Comment:
Consider adding `verbose` arg, and also having this function print out the info in a pretty format.

Reply:
Eg. I'd personally have this function organize this info in a pd.DataFrame something like this:
```
class1  class2  confusion_score
A          B           0.95
```
That can then easily be printed in nice format (and manipulated in various ways, which the current returned structure is harder to manipulate).

In future could consider plotting it as a heatmap, just like folks normally do with confusion matrix.

Reply:
For the top few most-confused class-pairs, it'd also be nice to print more info like:

cleanlab found 124 examples with given label A that should actually be labeled as class B

Reply:
My gut also was to return a dataframe. I didn't because right now Cleanlab doesn't depend on pandas. Can you think of any downsides to introducing pandas as a new dependency? I kept it to just numpy and small fast packages to keep Cleanlab lean as long as possible. Thoughts @anishathalye and @jwmueller ?

Reply:
I think it'll eventually have to support pandas in order to avoid having all sorts of arrays/lists all over tracking different data issues. Will probably add pandas dependency in 2.1 once multiple data issues are being flagged in a single run. And most data scientists will anyway already be using pandas (I've never personally had issue getting pandas to play nice with another package).  

For v2.0, I don't have strong preference, but it does seem like the return types are being somewhat sacrificed by not having pandas. And if we anyway want to add pandas dependency  in v2.1, then at that point would not want to be stuck with this suboptimal return type.

Reply:
Another option is to lazy import pandas inside each function with an informative error msg if it fails, so pandas is only strictly needed for user to utilize these `dataset` functions.

Reply:
made pandas a dependency for cleanlab. method now returns pd.DataFrame